### PR TITLE
DOC: add docs on using the `build-dir` config setting

### PR DESCRIPTION
**Pull request by rgommers**
Closes gh-540

Also touches on gh-246, which is a common enough hiccup that it's useful to explicitly show dealing with `meson-log.txt` in CI.

### CODE REVIEW

File: docs/how-to-guides/config-settings.rst [hunk 1]
```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
 
 .. _build: https://pypa-build.readthedocs.io/en/stable/
 .. _pip: https://pip.pypa.io/
+
+
+Using a persistent build directory
+==================================
+
+By default, ``meson-python`` will create a temporary build directory, which
+gets cleaned up after the sdist or wheel has been built, or if the build fails.
```
Comment (dnicolodi) [inline docs/how-to-guides/config-settings.rst:68]: I would tweak this to "By default, meson-python creates a temporary build directory, which is deleted when the build terminates." It is a bit more direct and avoid the expression "cleaned up" which is not well defined.

File: docs/how-to-guides/config-settings.rst [hunk 2]
```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
...
+The ``build-dir`` config seting (see :ref:`reference-config-settings`) can be
```
Comment (dnicolodi) [inline docs/how-to-guides/config-settings.rst:69]: I would simply link "config settings" to the relative documentation page.

File: docs/how-to-guides/config-settings.rst [hunk 3]
```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
...
+used to build in a user-specified build directory instead. For example:
```
Comment (dnicolodi) [inline docs/how-to-guides/config-settings.rst:70]: Should we say immediately here that the user-specified build directory is left alone? Maybe it is obvious.

File: docs/how-to-guides/config-settings.rst [hunk 4]
```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
...
+
+    .. tab-item:: pip
+        :sync: key_pip
+
+	.. code-block:: console
+
+	   $ python -m pip install . -Cbuild-dir=build
+
+
+The ``build`` directory will now contain all the build output created by
```
Comment (dnicolodi) [inline docs/how-to-guides/config-settings.rst:89]: I would say "build artifacts and support files" instead than "build output". I think we use the latter to indicate the console logs usually.

File: docs/how-to-guides/config-settings.rst [hunk 5]
```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
...
+- For accessing build logs and intermediate build outputs when debugging
+
+The same build directory can be used by subsequent invocations of
+``meson-python``. This avoids having to rebuild the whole project when testing
+changes during development. This will work with any config settings flags,
+including for cross compilation. The one thing to keep in mind is that changing
+the version of ``meson`` or ``meson-python`` in the build environment between
+separate builds in the same build directory is not supported (it often works,
+but may fail - and the error message cannot be guaranteed to be clear about the
+reason).
```
Comment (dnicolodi) [inline docs/how-to-guides/config-settings.rst:103]: I don't think this is the case anymore. We fixed meson-python so that `--reconfigure` is used when needed.

File: docs/how-to-guides/config-settings.rst [hunk 6]
```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
...
+
+Using ``build-dir`` makes it possible to access the log files that Meson
+creates. When a build fails, Meson's CLI logging will display the reason for
```
Comment (dnicolodi) [inline docs/how-to-guides/config-settings.rst:106]: "...to access the Meson's log and introspection files." Shorter and more precise.

File: docs/how-to-guides/config-settings.rst [hunk 7]
```diff
@@ -59,3 +59,63 @@ immediately install it, replace ``wheel`` with ``install``.  See the
...
+the failure in most cases. However, it is sometimes necessary to look at more
+detailed logs like ``meson-logs/meson-log.txt``. A common reason for this is
```
Comment (dnicolodi) [inline docs/how-to-guides/config-settings.rst:108]: I would make this more to the point and more precise. I would remove the sentence starting "When a build fails..." and I would reword the next sentence to something like: "To diagnose why a build fails at the project configuration stage, it is sometimes necessary to inspect the detailed Meson log in the ``meson-logs/meson-log.txt`` file in the build directory. The complete path the the log file is available in the command output when the build fails. For example, when dependency detection fails it is often..."

File: docs/reference/config-settings.rst [hunk 8]
```diff
@@ -23,10 +23,6 @@ them.
    exists and contains a valid Meson build directory setup, the
    project will be reconfigured using ``meson setup --reconfigure``.
 
-   The same build directory can be used by subsequent invocations of
-   ``meson-python``. This avoids having to rebuild the whole project
-   when testing changes during development.
-
```
Comment (dnicolodi) [inline docs/reference/config-settings.rst:29]: Why is this paragraph removed?

**Review by dnicolodi — CHANGES_REQUESTED** [review]
Thanks @rgommers It is a very useful addition to the docs. I have a few stylistic comment and a remark regarding a limitation that is described but I don't think is there.

File: docs/reference/config-settings.rst (hunk 8 above)
Comment (dnicolodi) [reply docs/reference/config-settings.rst:29]: I'm ok with removing it.

File: docs/how-to-guides/config-settings.rst (hunk 5 above)
Comment (rgommers) [reply docs/how-to-guides/config-settings.rst:103]: I'm not sure that `meson setup --reconfigure` is 100% the same as `rm -rf builddir && meson setup builddir` though. I recently ran into a very odd failure (https://github.com/scipy/scipy/issues/20433) which had to do with a pickled version of `meson` that is reloaded at build time.

The Meson docs don't say anything about this either way I believe, but my take-away from https://github.com/mesonbuild/meson/pull/12357#issuecomment-2050016590 was "if you upgrade Meson, no guarantees unless you delete the build dir". I could be wrong though - @eli-schwartz?

Comment (eli-schwartz) [reply docs/how-to-guides/config-settings.rst:103]: Reconfigure is *definitely* a section of the codebase covered by meson's explicit check for coredata version.

If meson detects that you've upgraded meson and the version of the coredata doesn't match, it essentially does a wipe and from-scratch configure, instead of the normal reconfigure logic. Alternatively, if you haven't asked to setup the project it isn't apropos to do so automatically just because the coredata is incompatible -- from memory, things like test/install will just print a "fatal error: xxx. Try running meson setup --wipe".

Reconfigure is indeed different from deleting the build directory. It will:
- keep dependency lookup caches instead of actually running pkg-config/cmake for dependency()
- keep incremental build outputs, since they will anyways be invalidated if the command line changes
- preserve current option states

File: docs/how-to-guides/config-settings.rst (hunk 6 above)
Comment (rgommers) [reply docs/how-to-guides/config-settings.rst]: adopted

File: docs/how-to-guides/config-settings.rst (hunk 5 above)
Comment (rgommers) [reply docs/how-to-guides/config-settings.rst]: thanks both. removed these sentences now.

File: docs/how-to-guides/config-settings.rst (hunk 4 above)
Comment (rgommers) [reply docs/how-to-guides/config-settings.rst]: adopted

File: docs/how-to-guides/config-settings.rst (hunk 3 above)
Comment (rgommers) [reply docs/how-to-guides/config-settings.rst]: good idea I think, added "which will not be deleted"

File: docs/how-to-guides/config-settings.rst (hunk 7 above)
Comment (rgommers) [reply docs/how-to-guides/config-settings.rst]: adopted

File: docs/reference/config-settings.rst (hunk 8 above)
Comment (rgommers) [reply docs/reference/config-settings.rst:29]: okay, resolving this then.

**Comment by rgommers** [comment]
Addressed all comments, I think this is ready now.

**Comment by rgommers** [comment]
Cirrus is not happy. Can be ignored for this PR, since its docs-only.

**Comment by dnicolodi** [comment]
@rgommers I find fixing prose via GitHub reviews not very efficient. I edited your text to make it flow better and pushed it to your branch. No conceptual changes intended. Please have a look.

**Comment by henryiii** [comment]
We could change the other places that have a pip version to use `-C` now, I think? (Not in this PR!)

**Comment by rgommers** [comment]
> We could change the other places that have a pip version to use `-C` now, I think? (Not in this PR!)

Agreed.

**Comment by rgommers** [comment]
The rephrasings LGTM - time to get this in. Thanks @dnicolodi and @henryiii
